google/resource_spanner_database.go

```diff
@@ -1,216 +1,255 @@
+// ----------------------------------------------------------------------------
+//
+//     ***     AUTO GENERATED CODE    ***    AUTO GENERATED CODE     ***
+//
+// ----------------------------------------------------------------------------
+//
+//     This file is automatically generated by Magic Modules and manual
+//     changes will be clobbered when the file is regenerated.
+//
+//     Please read more about how to change this file in
+//     .github/CONTRIBUTING.md.
+//
+// ----------------------------------------------------------------------------
+
 package google
 
 import (
 	"fmt"
-	"github.com/hashicorp/terraform/helper/schema"
 	"log"
-	"net/http"
-	"regexp"
+	"reflect"
+	"time"
 
-	"google.golang.org/api/googleapi"
-	"google.golang.org/api/spanner/v1"
-)
-
-const (
-	spannerDatabaseNameFormat = "^[a-z][a-z0-9_-]*[a-z0-9]$"
+	"github.com/hashicorp/terraform/helper/schema"
 )
 
 func resourceSpannerDatabase() *schema.Resource {
 	return &schema.Resource{
 		Create: resourceSpannerDatabaseCreate,
 		Read:   resourceSpannerDatabaseRead,
 		Delete: resourceSpannerDatabaseDelete,
+
 		Importer: &schema.ResourceImporter{
-			State: resourceSpannerDatabaseImport("name"),
+			State: resourceSpannerDatabaseImport,
+		},
+
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(240 * time.Second),
+			Delete: schema.DefaultTimeout(240 * time.Second),
 		},
 
 		Schema: map[string]*schema.Schema{
 			"instance": {
-				Type:     schema.TypeString,
-				Required: true,
-				ForceNew: true,
+				Type:             schema.TypeString,
+				Required:         true,
+				ForceNew:         true,
+				DiffSuppressFunc: compareSelfLinkOrResourceName,
 			},
-
 			"name": {
-				Type:         schema.TypeString,
-				Required:     true,
-				ForceNew:     true,
-				ValidateFunc: validateResourceSpannerDatabaseName,
```

Comment:
Should we keep this `ValidateFunc`?

Reply:
My philosophy on `ValidateFunc`s is that the bad _veeeeeeery slightly_ outweighs the good.

The good: we stop users from tearing down their infrastructure, beginning to recreate it, and finding a problem during recreation.  That situation sucks!  But, on the other hand, the user is only going to be in that situation if they changed the field in question, so you gotta imagine the debugging experience is pretty straightforward.

The bad: we end up maintaining `ValidateFunc`s as the actual requirements of the API shift, so users end up waiting for our next release to be able to import their existing OOB-created resources (this is mostly a problem for new regions - name requirements change less often).  There's also the maintenance burden on us - we don't have a good way to autogenerate the actual code of a ValidateFunc, so we end up maintaining it as custom code.

Our collective philosophy is really mixed on them - we include them some places and don't include them other places, even when all the same arguments apply in both places.

I'd be inclined to remove 'em all, but very mildly inclined.  What do you think?

Reply:
> user is only going to be in that situation if they changed the field in question, so you gotta imagine the debugging experience is pretty straightforward

I'm not sure that's a given- our tests are pretty simple, but a lot of users will be using much more complex TF configs, not only by size, but by including things like modules; fields changed aren't necessarily 1:1 with config line changes, and early warning signs *are* useful to them.

> this is mostly a problem for new regions - name requirements change less often

Yeah I am of the feeling that validating regions is a Bad Idea. We've mostly removed that I think, luckily.

> we don't have a good way to autogenerate the actual code of a ValidateFunc, so we end up maintaining it as custom code.

Mostly we just haven't written anything to autogenerate it; I bet we could introduce a lot of validation really fast with MM fan-out 🙂

---

I'm for keeping this validation, if we're not seeing _problems_ with it, someone clearly thought it was worth adding in the first place.

Reply:
Also:

> users end up waiting for our next release to be able to import their existing OOB-created resources 

I bet that's a solvable problem. What if users could opt out of validation? 3 ideas off the top of my head in order of how much I like them:

* HashiCorp could add a `lifecycle` flag to Terraform core to ignore validatefuncs on a resource
* I think apply-time interpolated values don't get validated; a dummy resource could trick Terraform into skipping validation, at the cost of a less intuitive config. We could write a canonical example of how to do so.
* We could add a `SKIP_GOOGLE_PROVIDER_VALIDATION` env var, and check for it being `true` in all our validatefuncs (this is mostly only feasible if we autogenerate a bunch of them)

Reply:
Oh hey, someone wrote `Provider::Terraform::Validation`!  This rocks.  Never mind, maintenance burden is almost trivial now.  Pushed up that change.